### PR TITLE
feat(nx-cloud): get git remote info to use in cloud connect

### PR DESCRIPTION
**Pull request by mandarini**
CLOSED in favor of: https://github.com/nrwl/nx/pull/26262

THIS PR IS THE FIRST IN A CHAIN OF PRs. The last PR in the chain is: https://github.com/nrwl/nx/pull/26262

So if https://github.com/nrwl/nx/pull/26262 is OK, then maybe we just merge https://github.com/nrwl/nx/pull/26262?

Get info from setup remotes (if any), identify if github, if github get github slug, to be able to generate GH cloud onboarding URL.

Todo:
- [x] hide behind feature flag
- [x] create commit message for PR
- [x] add PR instructions

Example screenshots from `nx init` and `nx connect`:

Screenshot 2024-05-29 at 4 15 19 PM

Screenshot 2024-05-29 at 4 15 36 PM

### CODE REVIEW

File: packages/nx/src/command-line/connect/connect-to-nx-cloud.ts [hunk 1]
```diff
@@ -51,12 +52,39 @@ export async function connectToNxCloudIfExplicitlyAsked(
...
+        `Unable to authenticate. Either define accessToken in nx.json or set the NX_CLOUD_ACCESS_TOKEN env variable.`
+      );
+    }
+
     output.log({
       title: '✔ This workspace already has Nx Cloud set up',
       bodyLines: [
         'If you have not done so already, connect your workspace to your Nx Cloud account:',
-        `- Login at ${getNxCloudUrl(nxJson)} to connect your repository`,
+        `- Do so by going to the following URL:
```
Comment (jaysoo) [inline packages/nx/src/command-line/connect/connect-to-nx-cloud.ts:85]: I'd reword this to be more terse.

`- Connect with Nx Cloud at ${connectCloudUrl}`

Just to reinforce what going to the URL will do.

File: packages/nx/src/nx-cloud/generators/connect-to-nx-cloud/connect-to-nx-cloud.ts [hunk 2]
```diff
@@ -72,18 +73,32 @@ async function createNxCloudWorkspace(
...
+  let connectCloudUrl;
+  if (githubSlug) {
+    connectCloudUrl = `${apiUrl}/setup/connect-workspace/vcs?provider=GITHUB&selectedRepositoryName=${encodeURIComponent(
+      githubSlug
+    )}`;
+  } else {
+    connectCloudUrl = `https://cloud.nx.app/setup/connect-workspace/manual?accessToken=${token}`;
+  }
   output.note({
     title: `Your Nx Cloud workspace is public`,
```
Comment (jaysoo) [inline packages/nx/src/nx-cloud/generators/connect-to-nx-cloud/connect-to-nx-cloud.ts:103]: We've received a lot of feedback that this message is kinda scary, and users aren't sure if they should be concerned: `Your Nx Cloud workspace is public`. 

Do we want to reword it while we're in here? Something like:

```
Your Nx Cloud workspace is ready

To claim it, connect it to your Nx Cloud account:
- Push your changes
- Connect your workspace here ${connectCloudUrl}
```

File: packages/nx/src/nx-cloud/utilities/git-info.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import { execSync } from 'child_process';
+
+export function getGithubSlugOrNull(): string | null {
+  const gitRemote = execSync('git remote -v').toString();
+  return extractUserAndRepoFromGitHubUrl(gitRemote);
+}
+
+function extractUserAndRepoFromGitHubUrl(gitRemotes: string): string | null {
```
Comment (jaysoo) [inline packages/nx/src/nx-cloud/utilities/git-info.ts:8]: A unit test to show this working would be good.

Comment (jaysoo) [reply packages/nx/src/nx-cloud/utilities/git-info.ts:8]: Also, if we only support `origin` as remote name, should it be `git config --get remote.origin.url`?

Comment (mandarini) [reply packages/nx/src/nx-cloud/utilities/git-info.ts:8]: We don't really only support origin. Here's the thing.

If a user has `origin` set up and the `origin` is a github URL, then we'll use that. If the user does not have a remote named `origin`, then we pick the first remote in the list that has a github URL. 

Even if we do pick `origin` or in the absence of `origin` we pick the first remote with github, it still does not guarantee that it's the remote they want to set up. they could be working with multiple forks, or custom named remotes, etc. So we either pick origin or just one, and if it's the wrong one, then they'll have to change the URL themselves.

We could list their remotes and tell them "pick the one you want" but I think it's an overkill?

Comment (mandarini) [reply packages/nx/src/nx-cloud/utilities/git-info.ts:8]: (or `git remote get-url origin` :P )

Comment (mandarini) [reply packages/nx/src/nx-cloud/utilities/git-info.ts:8]: I'm adding a unit test now!

File: packages/nx/src/nx-cloud/utilities/git-commit.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
+import * as chalk from 'chalk';
+import { logger } from '../../utils/logger';
+import { commitChanges } from '../../command-line/migrate/migrate';
```
Comment (FrozenPandaz) [inline packages/nx/src/nx-cloud/utilities/git-commit.ts:3]: Can we move this to a different file rather than import it from `migrate`?

File: packages/nx/src/command-line/migrate/migrate.ts [hunk 5]
```diff
@@ -1588,7 +1588,7 @@ function commitChanges(commitMessage: string): string | null {
   return getLatestCommitSha();
 }
 
-function getLatestCommitSha(): string | null {
+export function getLatestCommitSha(): string | null {
```
Comment (FrozenPandaz) [inline packages/nx/src/command-line/migrate/migrate.ts:1591]: I don't think this needs to be exported?

File: packages/nx/src/command-line/migrate/migrate.ts [hunk 6]
```diff
@@ -1573,7 +1573,7 @@ function getStringifiedPackageJsonDeps(root: string): string {
   }
 }
 
-function commitChanges(commitMessage: string): string | null {
+export function commitChanges(commitMessage: string): string | null {
```
Comment (FrozenPandaz) [inline packages/nx/src/command-line/migrate/migrate.ts:1576]: Move this into some nx utils for git.

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

File: packages/nx/src/command-line/migrate/migrate.ts (hunk 5 above)
Comment (mandarini) [reply packages/nx/src/command-line/migrate/migrate.ts:1591]: oooops!

File: packages/nx/src/utils/git-utils.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
+import { execSync } from 'child_process';
+
+export function getGithubSlugOrNull(): string | null {
+  const gitRemote = execSync('git remote -v').toString();
```
Comment (jdpearce) [inline packages/nx/src/utils/git-utils.ts:4]: Could we use `git ls-remote --get-url` instead? We wouldn't have to parse anything then.

File: packages/nx/src/utils/git-utils.spec.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
+import { extractUserAndRepoFromGitHubUrl } from './git-utils';
+
+describe('extractUserAndRepoFromGitHubUrl', () => {
```
Comment (jdpearce) [inline packages/nx/src/utils/git-utils.spec.ts:3]: What about the case where the remote starts with `https`?

File: packages/nx/src/nx-cloud/generators/connect-to-nx-cloud/connect-to-nx-cloud.ts [hunk 9]
```diff
@@ -72,20 +73,55 @@ async function createNxCloudWorkspace(
...
-  let origin = 'https://nx.app';
-  try {
-    origin = new URL(url).origin;
-  } catch (e) {}
+function printSuccessMessage(
+  url: string,
+  token: string,
+  installationSource: string
+) {
+  if (process.env.NX_NEW_CLOUD_ONBOARDING !== 'true') {
```
Comment (jdpearce) [inline packages/nx/src/nx-cloud/generators/connect-to-nx-cloud/connect-to-nx-cloud.ts:81]: What's the plan to remove this?

File: packages/nx/src/utils/git-utils.ts (hunk 7 above)
Comment (mandarini) [reply packages/nx/src/utils/git-utils.ts:4]: I want to make sure we're getting the right remote, this is why I'm listing all.

File: packages/nx/src/nx-cloud/generators/connect-to-nx-cloud/connect-to-nx-cloud.ts (hunk 9 above)
Comment (mandarini) [reply packages/nx/src/nx-cloud/generators/connect-to-nx-cloud/connect-to-nx-cloud.ts:81]: I have no idea!

Comment (mandarini) [reply packages/nx/src/nx-cloud/generators/connect-to-nx-cloud/connect-to-nx-cloud.ts:81]: I assume when we are ready to release the feature?
